core/src/main/scala/cats/Traverse.scala

```diff
@@ -135,12 +135,24 @@ import simulacrum.typeclass
   override def map[A, B](fa: F[A])(f: A => B): F[B] =
     traverse[Id, A, B](fa)(f)
 
+  /**
+   * Akin to [[map]], but allows to keep track of a state value
+   * when calling the function.
+   */
+  def mapAccumulate[A, S, B](fa: F[A])(init: S)(f: (A, S) => (B, S)): (F[B], S) =
```

Comment:
Actually, I think that parameters and results of the `f` function should be flipped.
```suggestion
  def mapAccumulate[A, S, B](fa: F[A])(init: S)(f: (S, A) => (S, B)): (S, F[B]) =
```
This order is used in [FS2](https://github.com/typelevel/fs2/blob/ff9d758ae61cb19ad4d6f5cde4d0b236efab8157/core/shared/src/main/scala/fs2/Stream.scala#L1770)
and it is usually used with states – S usually goes first in a pair.

Also I am not certain if the overall result should really be `(S, F[B])` rather than `F[(S, B)]`.

Reply:
Agree on `S` coming first.

I believe `(S, F[B])` is correct, I guess `F[(S, B)]` would also be storing the state for each iteration?

Reply:
Correct, and that would reflect the idea of multiple state transitions, I think: so we have a function that makes a single transition `(S, A) => (S, B)`, where `A` is an input event, and `B` is an output one associated with a particular state transition. Therefore, when we return `F[(S, B)]`, we get all states and outputs linked to each other (in context of `F[_]` of course).

But it is just IMO. And that is how it is defined in FS2 after all :)

Reply:
...on the second thought, the result could be just `F[B]`, to keep `mapAccumulate` consistent (to some extent) with `mapWithIndex` (which simply returns `F[B]`, not `(Int, F[B])`).

Reply:
Also I'm thinking about a possible next step: once we get `mapAccumulate` then someone may ask for `mapAccumulateA`:
```scala
def mapAccumulateA[A, S, B](fa: F[A])(init: S)(f: (A, S) => F[(B, S)]): RESULT_TYPE
```
So what would be `RESULT_TYPE` in that case?

Reply:
@satorg about the order of parameters, I was curious about what you all think since I couldn't think of a good argument for either order and I keep flipping them when asking if it existed and when proposing it on **scala-library-next**. I was sure I based this one on **fs2** but it seems I confused myself at the final moment; again 😅 
However, I also used this order because `mapWithIndex` has `(A, Int)` instead of `(Int, A)`, thus it seemed natural that the `A` was always the first parameter.
Nevertheless, I am okay with any order.

About returning `F[(S, B)]` _(or `F[(B, S)]`)_, I think it makes sense, I originally only care about the final `F[B]` then, I added the final state as suggested in the issue I opened in **scala-library-next**.

Reply:
I actually don't know why `mapWithIndex` uses this parameter order. Perhaps (I'm guessing) just to be in par with `zipWithIndex` which also has index as a second item in the tuple it returns. But it is fine to me since the index is not truly a state (although it can be modeled with the latter). But AFAI can see, in most other methods that work with states or accumulators the state parameter usually goes first (that said, let's assume that `foldLeft` is an example of such behavior while `foldRight` is an exception from that :) )

Regarding the result type. Yes, good point – having `F[B]` we can always make `F[(S, B)]` out of it. And having `F[(S, B)]` we can always get the last `S`. So seems that returning `F[B]` would be the most generic approach. And regarding the `(S, F[B])` the only "pros" I can see here – well, basically we already have the final `S` at the end of the calculations, why not to return it, right? On the other hand, there are a few of "cons" as well:
1. If some one still needs to get _all_ states, not just the final one, they will end up with something like `(S, F[(S, B)])` which is way too clumsy (to me at least).
2. Receiving some final value out of `Traverse` – it is a "folding" job, not a "mapping" one. I.e. since `Traverse` extends `Foldable` it is always possible to get just a final `S` by calling some kind of `fold` methods.
3. Returning `S` directly from a map-like methods looks like escaping of `F` context which seems a bit suspicious to me.

So yeah, as a summary I would say that my objections against `(S, F[B])` are more aesthetic rather than practical.

Reply:
I am very -1 on `F[(S, B)]`. That gratuitously puts S inside potentially and effect when it does not need to be. If F is IO that means we can no longer access S without evaluating the IO when in fact S is a function of the length of the traversable and not dependent on anything about the effect.

Reply:
But `Traverse` cannot be implemented for any effect, I guess. Or can it? I assume it is impossible just because it is also a `Foldable` which would mean unsafe direct access to the contained value.

Reply:
I mean, folding an effect is only feasible by triggering its evaluation.

Reply:
@satorg 

> But AFAI can see, in most other methods that work with states or accumulators the state parameter usually goes first

It makes sense. And, since I don't have any preference and I don't see any argument for the current order I will change it. Unless either @armanbilge or @johnynek may have something to say about it.

> we can always make F[(S, B)] out of it. 

Also `F[(B, S)]` or `F[(S, B, S)]` _(input and output state)_, thus yeah I don't think it is worth it to have that there. Also, if you don't care about the states after the `map` you would need to discard them; and I would think is more common to not care about them than care... but, of course, we lack any data for either affirmation.

> And having F[(S, B)] we can always get the last S

Yes, but it may require some _(inefficient)_ traversal of the collection.

> And regarding the (S, F[B]) the only "pros" I can see here – well, basically we already have the final S at the end of the calculations, why not to return it, right?

I mean, as I said, my original idea was also just only `F[B]` it was a comment on **scala-library-next** that proposed returning the last state, which makes sense.
However, being honest, I think it hurts operation chaining, sure you can call `_._1` or `_._2` to get the `F[B]`, but I personally never liked the tuple accessors they are too opaque.

Also, if someone wants the final state they can just reimplement the logic, is not a big deal anyway.
It may be considered again if many folks ask about a way to get the final state.

@johnynek 

> That gratuitously puts S inside potentially and effect when it does not need to be.
> If F is IO that means we can no longer access S

If `F` is `IO` this won't work because there is no `Traverse[IO]` :)
Also, this function only makes sense for `Fs` that actually contain values _(note the plural)_ inside of them; e.g. `List`, `Stream`. Rather than `Fs` that are programs descriptions; e.g. `IO`, `Option`, `Either`

Reply:
yeah, sorry, I got confused there (since this is a Traverse and the generally the interesting traversals are ones where you have both the `F[_]` of the container type a `G[_]` of Applicative type, but I digress, I was incorrect).

I still feel not putting it inside is right. Suppose I don't want it inside. Forcing it inside makes me have to map it away after. Suppose you do want it inside, then just use a `B1 = (B, S)` and put it inside as you like.

So, I think the better API is to keep it as it was originally `(F[B], S)` (or reversed order, I have less strong feeling about the order).

Reply:
But do we really need that `S` in the result eventually? Again, comparing `mapAccumulate` to `mapWithIndex` – the latter does not return `(F[B], Int)`, but rather just `F[B]`.

I'm a bit reluctant just because I don't think there will be a lot of use cases where both `S` and `F[B]` will be requested for. I would suppose that it will be either of them but not both. And if only `S` is necessary, then `foldLeft` will be a better method to call.

The only reason to return both I can really see – these values both are already in place. So we just take everything we have and return it in a tuple. And I admit – it is not necessarily the illicit reason though...

Reply:
Let me try one more argument:

If we change the order of the arguments and return a function:
```scala
def mapAccumulate[A, S, B](fn: (A, S) => (B, S)): (F[A], S) => (F[B], S) = ...
```
so we see the symmetry here: map-accumulate is lifting a function of type `(A, S) => (B, S)` to a function of type `(F[A], S) => (F[B], S)`.

This can be compared to others:
```scala
def map[A, B](fn: A => B): F[A] => F[B] = ..
def flatMap[A, B](fn: A => F[B]): F[A] => F[B] = ..
def map2[A, B, C](fn: (A, B) => C): (F[A], F[B]) => F[C] = ...
def tailRecM[A, B](fn:  A => F[Either[A, B]): A => F[B]

def foldLeft[A, B](fn: (A, B) => A): (A, F[B]) => A = ...
def traverse[A, B, G[_]](fn: A => G[B]): F[A] => G[F[B]]  = ...

// and now:
def mapAccumulate[A, S, B](fn: (A, S) => (B, S)): (F[A], S) => (F[B], S) = ...
```

it's really helpful, I think, to think of these functions convert on function to a very similar function with type constructors moved around.

Reply:
Oh, this is a cool one, I really like that point of view.